crates/hir_ty/src/checker/mutability.rs

```diff
@@ -0,0 +1,189 @@
+            hir::Stmt::Item { item } => match item {
+                hir::Item::Function(_) => (),
+                hir::Item::Module(_) => (),
+                hir::Item::UseItem(_) => (),
+            },
+        }
+    }
+
+    fn check_expr(&mut self, expr_id: hir::ExprId) {
+        let expr = expr_id.lookup(self.hir_file.db(self.db));
+        match expr {
+            hir::Expr::Block(block) => {
+                for stmt in &block.stmts {
+                    self.check_stmt(stmt);
+                }
+                if let Some(tail) = block.tail {
+                    self.check_expr(tail);
+                }
+            }
+            hir::Expr::If {
+                condition,
+                then_branch,
+                else_branch,
+            } => {
+                self.check_expr(*condition);
+                self.check_expr(*then_branch);
+                if let Some(else_branch) = else_branch {
+                    self.check_expr(*else_branch);
+                }
+            }
+            hir::Expr::Symbol(symbol) => match symbol {
+                hir::Symbol::Param { .. } => (),
+                // `VariableDef`で処理済みのため、ここでは何もしません
+                hir::Symbol::Local { .. } => (),
+                hir::Symbol::Missing { .. } => (),
+            },
+            hir::Expr::Binary { op, lhs, rhs } => {
+                self.check_expr(*lhs);
+                self.check_expr(*rhs);
+
+                let lhs = lhs.lookup(self.hir_file.db(self.db));
+                if let hir::Expr::Symbol(hir::Symbol::Local { binding, .. }) = lhs {
+                    let binding = binding.lookup(self.hir_file.db(self.db));
+                    if *op == hir::BinaryOp::Assign && !binding.mutable {
+                        self.errors
+                            .push(TypeCheckError::ImmutableReassignment { expr: expr_id });
+                    }
+                }
+            }
+            hir::Expr::Literal(_) => (),
+            hir::Expr::Unary { op: _, expr } => {
+                self.check_expr(*expr);
+            }
+            hir::Expr::Call { callee, args } => {
+                for arg in args {
+                    self.check_expr(*arg);
+                }
+
+                match callee {
+                    hir::Symbol::Local { .. } => (),
+                    hir::Symbol::Param { .. } => (),
+                    hir::Symbol::Missing { .. } => (),
+                }
+            }
+            hir::Expr::Return { value } => {
+                if let Some(value) = value {
+                    self.check_expr(*value);
+                }
+            }
+            hir::Expr::Loop { block } => {
+                self.check_expr(*block);
+            }
+            hir::Expr::Continue => (),
+            hir::Expr::Break { value } => {
+                if let Some(value) = value {
+                    self.check_expr(*value);
+                }
+            }
+            hir::Expr::Missing => (),
+        }
```

Comment:
**Optimization Issue**: The method 'check_expr' performs multiple lookups and checks in a single function, which can be optimized by separating concerns and caching frequently accessed data. <br> **Fix**: Refactor 'check_expr' into smaller, focused functions and implement caching for database lookups to enhance performance. <br> **Code Suggestion**: 
 ```
 - let expr = expr_id.lookup(self.hir_file.db(self.db));
 + let expr = self.cached_lookup(expr_id);
 ```